### DISCUSSION TITLE
Read functions of HttpResponseStream are "NotSupported"

### DISCUSSION THREAD

**Discussion by waterlx** [dotnet/aspnetcore#57520] [Q&A]
We are using ASP .NET core 6.0.33 and woud like to read from ControllerBase.Response.Body (or ControllerBase.HttpContext.Response.Body) , which is a Stream. It seems any further operation on that Stream will fail and we got "Stream error in the HTTP/2 framing layer" as the response.

The actual type of that Stream is HttpResponseStream and the [source code](https://github.com/dotnet/aspnetcore/blob/63c492e22b06d4903cb4e7aee037295c71e1ec37/src/Servers/Kestrel/Core/src/Internal/Http/HttpResponseStream.cs#L40C1-L47C44) shows that all its read functions are "Not Supported".  And "CanRead" is also "false". And we guess those leads to that error response above.

Could anyone share your insights on these questions:
1. What is this "HttpResponseStream" used for?
2. Why is "read" not supported in this class?
3. Our existing code uses HttpContent.CopyToAsync() or Stream.CopyToAsync() to write to ControllerBase.Response.Body, which is a HttpResponseStream. Are they good practices?
4. The framework seems could handle HttpResponseStream well and make it as a good response with no errors, if we do not try to operate HttpResponseStream further and return that directly in an API of a controller. How does the framework make it?
5. If we have to read from HttpResponseStream, how could we do that?

**Comment by chrisoverzero** [comment]
What would you expect reading from the response stream to do? The response stream is for writing bytes back to the requester; it “gets read” far away over the network. It also can’t be seeked (sought?) for the same reason. You can’t un-transmit bytes.

**Reply by waterlx** [reply]
@chrisoverzero Thanks for helping out!

We have some legacy code doing the following: 
An existing API A internally calls some other APIs to get their responses, which is either Stream or HttpResponseMessage.
- For Stream, it calls Stream.CopyToAsync(ControllerBase.Response.Body) to write into ControllerBase.Response.Body
- For HttpResponseMessage, it gets Content (as HttpContent) from HttpResponseMessage firstly, and then calls HttpContent.CopyToAsync(ControllerBase.Response.Body) to write into ControllerBase.Response.Body.

So eventually I have ControllerBase.Response.Body (as HttpResponseStream) to contain the payload I need.

Previously, API A is called by ASP.NET framework, and I can have the correct response (I wonder why and how it could read something out of HttpResponseStream, which does not support "read")
Currently, I am writing a new API B, which needs to call API A and then does some modification then return. And I failed to do that. The error message is "Error: Stream error in the HTTP/2 framing layer"

**Reply by BrennanConroy** [reply]
The simplest thing would be to make API B pass in a stream to API A that is both writable and readable that way API B can read the "response" from API A.

e.g. pseudo code:
```
void ApiA()
{
    ApiAInner(ControllerBase.Response.Body);
}

void ApiB()
{
    var stream = new MemoryStream();
    ApiAInner(stream);

    // stream can be read here and then any modifications can be done
    stream.CopyToAsync(ControllerBase.Response.Body);
}

void ApiAInner(Stream response)
{
    response.Write(...);
}
```
